### ISSUE TITLE
Macro annotations cause "template" keyword to not be highlighted

### ISSUE THREAD

**Issue by HighCommander4** [jeff-hykin/cpp-textmate-grammar#457]
In the following code:

```c++
#  define ANNOTATION1
#  define ANNOTATION2

struct ANNOTATION1 ANNOTATION2 Waldo {
  template <typename T>   // <-- "template" keyword not highlighted
  void method(T);
};
```

The `template` keyword on the indicated line is not highlighted.

(The "annotations" are just macro names with empty expansions which are used by a static analysis tool. Trigger the bug seems to require two of them.)

**Comment by matter123** [comment]
This is generally unsolvable with the way that vscode currently uses for syntax highlighting. If the annotations are for a particularly popular static analysis tool, the annotations could be hardcoded.

**Comment by HighCommander4** [comment]
I brought this up [earlier](https://github.com/jeff-hykin/cpp-textmate-grammar/issues/371#issuecomment-524648443), but I'd just like to probe a bit further...

While I understand that keywords like `template` are useful for making inferences about the tokens that follow it, is it not possible to add "another pass" which just colorizes every occurrence of the `template` token?

**Comment by matter123** [comment]
Textmate grammars are a single pass, however, it might be possible to use an injection grammar to be able to perform another pass.

**Comment by matter123** [comment]
![Screen Shot 2020-03-03 at 12 14 02](https://user-images.githubusercontent.com/714007/75815648-a413a200-5d48-11ea-927e-ee962a74257e.png)

This is a simple injection grammar that highlights every keyword in blue. This does seem to allow for performing another pass, however it is far from a complete solution as the injection grammar still needs to have some sort of context-sensitivity, in order to avoid highlighting keywords in comments.

**Comment by jeff-hykin** [comment]
@HighCommander4  In general stuff like this with macros can't be supported because of things like

```
#define something1  myFunction(int a)
#define something2 const

struct MyStruct something1 something2 {
     template <typename T>  // <- this is invalid because these brackets {} are for a function not a struct
};
```
However, this one case I was able to push a "fix" with `v1.14.28`


@matter123  I optimized the struct/class block finder and assumed all preceding identifiers are macros since this appears to be a somewhat common practice, and because this syntax would otherwise be invalid. I made sure to handle the edgecases of `final` and the class name.
